src/iterative_ensemble_smoother/experimental_sies.py

```diff
@@ -0,0 +1,315 @@
+from __future__ import annotations
+from typing import Optional, TYPE_CHECKING, Callable
+
+import numpy as np
+import scipy as sp
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+
+# =============================================================================
+# MAIN CLASS
+# =============================================================================
+
+
+class SIES:
+    def __init__(
+        self,
+        param_ensemble: npt.NDArray[np.double],
+        observation_errors: npt.NDArray[np.double],
+        observation_values: npt.NDArray[np.double],
+        *,
+        seed: Optional[int] = None,
+    ):
+        self.iteration = 1
+        self.rng = np.random.default_rng(seed)
+        self.X = param_ensemble
+        self.d = observation_values
+        self.C_dd = observation_errors
+        self.A = (self.X - self.X.mean(axis=1, keepdims=True)) / np.sqrt(
+            self.X.shape[1] - 1
+        )
+
+        if self.C_dd.ndim == 2:
+            self.C_dd_cholesky = sp.linalg.cholesky(
+                self.C_dd,
+                lower=True,
+                overwrite_a=False,
+                check_finite=True,
+            )
+        else:
+            self.C_dd_cholesky = np.sqrt(self.C_dd)
+
+        # Equation (14)
+        self.D = self.d.reshape(-1, 1) + sample_mvnormal(
+            C_dd_cholesky=self.C_dd_cholesky, rng=self.rng, size=self.X.shape[1]
+        )
+
+        self.W = np.zeros(shape=(self.X.shape[1], self.X.shape[1]))
+        self.X_i = self.X
+
+    def newton(self, Y, step_length=0.5):
+        """Implementation of Algorithm 1."""
+        g_X = Y.copy()
+
+        # Get shapes. Same notation as used in the paper.
+        N = Y.shape[1]  # Ensemble members
+        n = self.X.shape[0]  # Parameters (inputs)
+        m = self.C_dd.shape[0]  # Responses (outputs)
+
+        # Line 4 in Algorithm 1
+        Y = (g_X - g_X.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)
+
+        # Line 5
+        Omega = self.W.copy()
+        Omega -= Omega.mean(axis=1, keepdims=True)
+        Omega /= np.sqrt(N - 1)
+        Omega.flat[:: Omega.shape[0] + 1] += 1  # Add identity in place
+
+        # Line 6
+        if n < N - 1:
+            # There are fewer parameters than realizations. This means that the
+            # system of equations is overdetermined, and we must solve a least
+            # squares problem.
+
+            # An alternative approach to producing A_i would be keeping the
+            # returned value from the previous Newton iteration (call it X_i),
+            # then computing:
+            # A_i = (self.X_i - self.X_i.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)
```

Comment:
Perhaps we have equivalence, and we do not need to flag anything?
So it becomes a question of potential extra computation (if we could have used the final line directly) in an edge case, versus the cost of an additional parameter upon initialization (it could be optional).
<img width="332" alt="image" src="https://github.com/equinor/iterative_ensemble_smoother/assets/13650516/bb33e88e-e630-4bce-b83b-8005291a0fc7">

Reply:
Agreed that it would be present upon initialization. Agree that this is a trade-off vs extra computation or extra argument.

I think we should not have the argument, since it's fair to assume that if one is running ensemble smoothers one does not have a linear model. In the toy cases where the model is actually linear, we can just do one extra projection $A_i^+ A_i = I$. It's not a bottleneck.